pgwire: v25.1.7: FormatBinary is not supported by pg_vector and a couple other types

This issue was auto filed by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry Link: [https://cockroach-labs.sentry.io/issues/6659818515/?referrer=webhooks_plugin](https://cockroach-labs.sentry.io/issues/6659818515/?referrer=webhooks_plugin)

Panic Message:

```
encoding.go:892: error in argument for $3: unsupported OID 90006 with format code FormatBinary
(1) ×
Wraps: (2) ×
Wraps: (3) candidate pg code: 08P01
Wraps: (4) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind.func2
  | 	pkg/sql/conn_executor_prepare.go:507
  | [...repeated from below...]
Wraps: (5) error in argument for $3
Wraps: (6) assertion failure
Wraps: (7) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase.DecodeDatum
  | 	pkg/sql/pgwire/pgwirebase/encoding.go:892
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind.func2
  | 	pkg/sql/conn_executor_prepare.go:498
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execBind
  | 	pkg/sql/conn_executor_prepare.go:516
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
  | 	pkg/sql/conn_executor.go:2528
  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
  | 	pkg/sql/conn_executor.go:2306
  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
  | 	pkg/sql/conn_executor.go:1002
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands
  | 	pkg/sql/pgwire/conn.go:252
  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func4
  | 	pkg/sql/pgwire/server.go:1197
  | runtime.goexit
  | 	src/runtime/asm_amd64.s:1695
Wraps: (8) unsupported OID 90006 with format code FormatBinary
Error types: (1) *hintdetail.withDetail (2) *hintdetail.withDetail (3) *pgerror.withCandidateCode (4) *withstack.withStack (5) *errutil.withPrefix (6) *assert.withAssertionFailure (7) *withstack.withStack (8) *errutil.leafError
-- report composition:
*errutil.leafError: unsupported OID 90006 with format code FormatBinary
encoding.go:892: *withstack.withStack (top exception)
*assert.withAssertionFailure
*errutil.withPrefix: error in argument for $3
conn_executor_prepare.go:507: *withstack.withStack (1)
*pgerror.withCandidateCode: 08P01
*hintdetail.withDetail
*hintdetail.withDetail
(check the extra data payloads)
```

<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

[src/runtime/asm_amd64.s#L1694-L1696](src/runtime/asm_amd64.s#L1694-L1696)
[pkg/sql/pgwire/server.go#L1196-L1198](pkg/sql/pgwire/server.go#L1196-L1198)
[pkg/sql/pgwire/conn.go#L251-L253](pkg/sql/pgwire/conn.go#L251-L253)
[pkg/sql/conn_executor.go#L1001-L1003](pkg/sql/conn_executor.go#L1001-L1003)
[pkg/sql/conn_executor.go#L2305-L2307](pkg/sql/conn_executor.go#L2305-L2307)
[pkg/sql/conn_executor.go#L2527-L2529](pkg/sql/conn_executor.go#L2527-L2529)
[pkg/sql/conn_executor_prepare.go#L515-L517](pkg/sql/conn_executor_prepare.go#L515-L517)
[pkg/sql/conn_executor_prepare.go#L506-L508](pkg/sql/conn_executor_prepare.go#L506-L508)
[src/runtime/asm_amd64.s#L1694-L1696](src/runtime/asm_amd64.s#L1694-L1696)
[pkg/sql/pgwire/server.go#L1196-L1198](pkg/sql/pgwire/server.go#L1196-L1198)
[pkg/sql/pgwire/conn.go#L251-L253](pkg/sql/pgwire/conn.go#L251-L253)
[pkg/sql/conn_executor.go#L1001-L1003](pkg/sql/conn_executor.go#L1001-L1003)
[pkg/sql/conn_executor.go#L2305-L2307](pkg/sql/conn_executor.go#L2305-L2307)
[pkg/sql/conn_executor.go#L2527-L2529](pkg/sql/conn_executor.go#L2527-L2529)
[pkg/sql/conn_executor_prepare.go#L515-L517](pkg/sql/conn_executor_prepare.go#L515-L517)
[pkg/sql/conn_executor_prepare.go#L497-L499](pkg/sql/conn_executor_prepare.go#L497-L499)
[pkg/sql/pgwire/pgwirebase/encoding.go#L891-L893](pkg/sql/pgwire/pgwirebase/encoding.go#L891-L893)
</details>


```
src/runtime/asm_amd64.s in runtime.goexit at line 1695
pkg/sql/pgwire/server.go in pkg/sql/pgwire.(*Server).serveImpl.func4 at line 1197
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommands at line 252
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 1002
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 2306
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 2528
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind at line 516
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind.func2 at line 507
src/runtime/asm_amd64.s in runtime.goexit at line 1695
pkg/sql/pgwire/server.go in pkg/sql/pgwire.(*Server).serveImpl.func4 at line 1197
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommands at line 252
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 1002
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 2306
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 2528
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind at line 516
pkg/sql/conn_executor_prepare.go in pkg/sql.(*connExecutor).execBind.func2 at line 498
pkg/sql/pgwire/pgwirebase/encoding.go in pkg/sql/pgwire/pgwirebase.DecodeDatum at line 892
```

### Tags

| Tag | Value |
| --- | --- |
| Command | mt start-sql |
| Environment | v25.1.7 |
| Go Version | go1.22.8 X:nocoverageredesign |
| Platform | linux amd64 |
| Distribution | CCL |
| Cockroach Release | v25.1.7 |
| Cockroach SHA | 8e7ab9c1c4ec161768dd3c43133200e6039503c1 |
| # of CPUs | 32 |
| # of Goroutines | 380 |



Jira issue: CRDB-51280

--- a/pkg/sql/pgwire/pgwirebase/BUILD.bazel
+++ b/pkg/sql/pgwire/pgwirebase/BUILD.bazel
@@ -20,6 +20,7 @@ go_library(
     visibility = ["//visibility:public"],
     deps = [
         "//pkg/geo",
+        "//pkg/geo/geopb",
         "//pkg/settings",
         "//pkg/sql/catalog/colinfo",
         "//pkg/sql/lex",

--- a/pkg/sql/pgwire/pgwirebase/encoding.go
+++ b/pkg/sql/pgwire/pgwirebase/encoding.go
@@ -18,6 +18,7 @@ import (
 	"unicode/utf8"
 
 	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/cockroach/pkg/geo/geopb"
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/sql/lex"
 	"github.com/cockroachdb/cockroach/pkg/sql/oidext"
@@ -817,6 +818,43 @@ func DecodeDatum(
 				return nil, err
 			}
 			return tree.NewDTSVector(ret), nil
+		case oidext.T_pgvector:
+			// PG binary format is
+			//   2 bytes for dimensions
+			//   2 bytes for unused, and
+			//   4 bytes for each float4.
+			if len(b) < 4 {
+				return nil, pgerror.Newf(pgcode.Syntax, "vector requires at least 4 bytes for binary format")
+			}
+			dim := int(binary.BigEndian.Uint16(b))
+			b = b[4:]
+			if dim > vector.MaxDim {
+				return nil, vector.MaxDimExceededErr
+			}
+			if len(b) < 4*dim {
+				return nil, pgerror.Newf(pgcode.Syntax, "vector with %d dimensions requires %d bytes for binary format", dim, 4*dim)
+			}
+			v := make(vector.T, dim)
+			for i := 0; i < dim; i++ {
+				v[i] = math.Float32frombits(binary.BigEndian.Uint32(b))
+				b = b[4:]
+			}
+			return tree.NewDPGVector(v), nil
+		case oidext.T_box2d:
+			// Expect 8 bytes for each of LoX, HiX, LoY, HiY.
+			if len(b) < 32 {
+				return nil, pgerror.Newf(pgcode.Syntax, "box2d requires at least 32 bytes for binary format")
+			}
+			loX := math.Float64frombits(binary.BigEndian.Uint64(b[0:8]))
+			hiX := math.Float64frombits(binary.BigEndian.Uint64(b[8:16]))
+			loY := math.Float64frombits(binary.BigEndian.Uint64(b[16:24]))
+			hiY := math.Float64frombits(binary.BigEndian.Uint64(b[24:32]))
+			box := geo.CartesianBoundingBox{
+				BoundingBox: geopb.BoundingBox{
+					LoX: loX, HiX: hiX, LoY: loY, HiY: hiY,
+				},
+			}
+			return da.NewDBox2D(tree.DBox2D{CartesianBoundingBox: box}), nil
 		case oidext.T_geometry:
 			v, err := geo.ParseGeometryFromEWKB(b)
 			if err != nil {

--- a/pkg/sql/pgwire/testdata/pgtest/box2d
+++ b/pkg/sql/pgwire/testdata/pgtest/box2d
@@ -1,0 +1,32 @@
+# "ResultFormatCodes": [1] = binary
+send
+Parse {"Name": "s", "Query": "SELECT 'BOX(1 0,1 0)'::BOX2D;"}
+Bind {"DestinationPortal": "p", "PreparedStatement": "s", "ResultFormatCodes": [1]}
+Execute {"Portal": "p"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"binary":"3ff00000000000003ff000000000000000000000000000000000000000000000"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+send
+Parse {"Query": "SELECT $1::BOX2D"}
+Bind {"ParameterFormatCodes": [1], "Parameters": [{"binary":"3ff00000000000003ff000000000000000000000000000000000000000000000"}]}
+Execute
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"BOX(1 0,1 0)"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}

--- a/pkg/sql/pgwire/testdata/pgtest/pgvector
+++ b/pkg/sql/pgwire/testdata/pgtest/pgvector
@@ -45,3 +45,19 @@ ReadyForQuery
 {"Type":"DataRow","Values":[{"binary":"000200004000000040400000"}]}
 {"Type":"CommandComplete","CommandTag":"SELECT 2"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
+
+send
+Parse {"Query": "SELECT $1::VECTOR"}
+Bind {"ParameterFormatCodes": [1], "Parameters": [{"binary":"000100003f800000"}]}
+Execute
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"[1]"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}

--- a/pkg/util/vector/vector.go
+++ b/pkg/util/vector/vector.go
@@ -21,6 +21,8 @@ import (
 // MaxDim is the maximum number of dimensions a vector can have.
 const MaxDim = 16000
 
+var MaxDimExceededErr = pgerror.Newf(pgcode.ProgramLimitExceeded, "vector cannot have more than %d dimensions", MaxDim)
+
 // T is the type of a PGVector-like vector.
 type T []float32
 
@@ -38,7 +40,7 @@ func ParseVector(input string) (T, error) {
 	parts := strings.Split(input, ",")
 
 	if len(parts) > MaxDim {
-		return T{}, pgerror.Newf(pgcode.ProgramLimitExceeded, "vector cannot have more than %d dimensions", MaxDim)
+		return T{}, MaxDimExceededErr
 	}
 
 	vector := make([]float32, len(parts))

